addon/webextension/background/auth.js

```diff
@@ -0,0 +1,208 @@
+/* globals browser */
+/* globals main, makeUuid, deviceInfo, analytics, catcher, defaultSentryDsn, communication */
+
+window.auth = (function () {
+  let exports = {};
+
+  let registrationInfo;
+  let initialized = false;
+  let authHeader = null;
+  let sentryPublicDSN = null;
+  let abTests = {};
+
+  catcher.watchPromise(browser.storage.local.get(["registrationInfo", "abTests"]).then((result) => {
+    if (result.abTests) {
+      abTests = result.abTests;
+    }
+    if (result.registrationInfo) {
+      registrationInfo = result.registrationInfo;
+    } else {
+      registrationInfo = generateRegistrationInfo();
+      console.info("Generating new device authentication ID", registrationInfo);
+      return browser.storage.local.set({registrationInfo});
+    }
+  }));
+
+  exports.getDeviceId = function () {
+    return registrationInfo && registrationInfo.deviceId;
+  };
+
+  function generateRegistrationInfo() {
+    let info = {
+      deviceId: "anon" + makeUuid() + "",
+      secret: makeUuid()+"",
+      registered: false
+    };
+    return info;
+  }
+
+  function register() {
+    return new Promise((resolve, reject) => {
+      let registerUrl = main.getBackend() + "/api/register";
+      let req = new XMLHttpRequest();
+      req.open("POST", registerUrl);
+      req.setRequestHeader("content-type", "application/x-www-form-urlencoded");
+      req.onload = catcher.watchFunction(() => {
+        if (req.status == 200) {
+          console.info("Registered login");
+          initialized = true;
+          saveAuthInfo(JSON.parse(req.responseText));
+          resolve(true);
+          analytics.sendEvent("registered");
+        } else {
+          console.warn("Error in response:", req.responseText);
+          reject(new Error("Bad response: " + req.status));
+        }
+      });
+      req.send(uriEncode({
+        deviceId: registrationInfo.deviceId,
+        secret: registrationInfo.secret,
+        deviceInfo: JSON.stringify(deviceInfo())
+      }));
+    });
+  }
+
+  function login(options) {
+    let { ownershipCheck, noRegister } = options || {};
+    return new Promise((resolve, reject) => {
+      let loginUrl = main.getBackend() + "/api/login";
```

Comment:
We have a plan in the future to associate devices with identities, but this per-device automatic identity let's us onboard people with no gateways.  Probably the identity association will be done entirely through the website, and the add-on will only ever know about its own identity.

Though there was an idea of using sync, and syncing all these per-device identities across all devices.  The unpairing gets a little complicated in that case, as does keeping the server in sync with the add-on.